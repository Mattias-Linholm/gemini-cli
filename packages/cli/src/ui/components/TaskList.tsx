/**
 * @license
 * Copyright 2025 Google LLC
 * SPDX-License-Identifier: Apache-2.0
 */

import React, { useState, useEffect } from 'react';
import { Box, Text } from 'ink';
import { taskListState, Task } from '@google/gemini-cli-core';

export const TaskList: React.FC = () => {
  const [tasks, setTasks] = useState<Task[]>([]);

  useEffect(() => {
    const handleTasksUpdated = (updatedTasks: Task[]) => {
      console.log('TaskList: handleTasksUpdated received', updatedTasks);
      setTasks(updatedTasks);
    };

    taskListState.on('tasks_updated', handleTasksUpdated);

    // Initial load
    setTasks(taskListState.getTasks());

    return () => {
      taskListState.off('tasks_updated', handleTasksUpdated);
    };
  }, []);

  if (tasks.length === 0) {
    return null;
  }

  return (
    <Box flexDirection="column">
      <Text bold>Tasks:</Text>
      {tasks.map((task) => (
        <Text key={task.id}>
          {task.status === 'completed' ? (
            <Text color="green">[âœ”] {task.description}</Text>
          ) : task.status === 'in_progress' ? (
            <Text color="yellow">
              {'[>]'} {task.description}
            </Text>
          ) : (
            <Text>[ ] {task.description}</Text>
          )}
        </Text>
      ))}
    </Box>
  );
};
