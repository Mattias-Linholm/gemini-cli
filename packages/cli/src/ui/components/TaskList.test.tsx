/**
 * @license
 * Copyright 2025 Google LLC
 * SPDX-License-Identifier: Apache-2.0
 */

/**
 * @license
 * Copyright 2025 Google LLC
 * SPDX-License-Identifier: Apache-2.0
 */

import { render } from 'ink-testing-library';
import { TaskList } from './TaskList';
import { taskListState } from '@google/gemini-cli-core';
import { vi } from 'vitest';

describe('TaskList', () => {
  beforeEach(() => {
    vi.spyOn(taskListState, 'getTasks').mockReturnValue([]);
    vi.spyOn(taskListState, 'on');
    vi.spyOn(taskListState, 'off');
  });

  afterEach(() => {
    vi.restoreAllMocks();
  });

  it('should not render if no tasks are present', () => {
    const { lastFrame } = render(<TaskList />);
    expect(lastFrame()).toBe('');
  });

  it('should render pending tasks correctly', () => {
    vi.spyOn(taskListState, 'getTasks').mockReturnValue([
      { id: '1', description: 'Task 1', status: 'pending' },
      { id: '2', description: 'Task 2', status: 'pending' },
    ]);
    const { lastFrame } = render(<TaskList />);
    expect(lastFrame()).toMatchSnapshot();
  });

  it('should render in-progress tasks correctly', () => {
    vi.spyOn(taskListState, 'getTasks').mockReturnValue([
      { id: '1', description: 'Task 1', status: 'in_progress' },
    ]);
    const { lastFrame } = render(<TaskList />);
    expect(lastFrame()).toMatchSnapshot();
  });

  it('should render completed tasks correctly', () => {
    vi.spyOn(taskListState, 'getTasks').mockReturnValue([
      { id: '1', description: 'Task 1', status: 'completed' },
    ]);
    const { lastFrame } = render(<TaskList />);
    expect(lastFrame()).toMatchSnapshot();
  });

  it('should update when tasks_updated event is emitted', () => {
    const { lastFrame } = render(<TaskList />);
    expect(lastFrame()).toBe('');

    taskListState.emit('tasks_updated', [
      { id: '1', description: 'New Task', status: 'pending' },
    ]);

    expect(lastFrame()).toMatchSnapshot();
  });

  it('should clean up event listener on unmount', () => {
    const { unmount } = render(<TaskList />);
    expect(taskListState.on).toHaveBeenCalledWith(
      'tasks_updated',
      expect.any(Function),
    );
    unmount();
    expect(taskListState.off).toHaveBeenCalledWith(
      'tasks_updated',
      expect.any(Function),
    );
  });
});
