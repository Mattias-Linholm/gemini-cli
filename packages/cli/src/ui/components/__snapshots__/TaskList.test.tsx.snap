// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`TaskList > should render completed tasks correctly 1`] = `""`;

exports[`TaskList > should render in-progress tasks correctly 1`] = `""`;

exports[`TaskList > should render pending tasks correctly 1`] = `""`;

exports[`TaskList > should update when tasks_updated event is emitted 1`] = `""`;
